### PR TITLE
Streamline Orca Gucs

### PR DESCRIPTION
**Pull request by vraghavan78**
Similar to https://github.com/greenplum-db/gporca/pull/174 I have tried to 

* Enable analyzing root partitions
* Ensure that the name of the guc is clear
* Remove double negation (where possible)
* Update comments
* Co-locate gucs that have similar purpose
* Remove dead gucs
* Classify them correctly so that they are no longer hidden

I did these changes over couple days, so the commits are little scatter brained, my apologies.

### CODE REVIEW

**Comment by asimrp** [comment]
Can all the commits be squashed as one commit with a message summarizing the changes?

Some GUCs changed from developer-only to user-visible.  That means documentation needs to be updated accordingly.  Is that right?

**Comment by vraghavan78** [comment]
Yes. when I push it i will squash it into one commit.

Doc needs to change, Mel is aware of it. I am working on what needs to
change for docs.
ᐧ

**Comment by mkiyama** [comment]
How do I know which gucs are becoming end-user gucs?
Is it changing DEVELOPER_OPTIONS to \<something else\>
or is it more complicated.

**Comment by vraghavan78** [comment]
@mkiyama you are correct, DEVELOPER_OPTIONS to <something else> are the gucs that need to be documented.

File: src/backend/utils/misc/guc_gp.c [hunk 1]
```diff
@@ -2909,7 +2909,7 @@ struct config_bool ConfigureNamesBool_gp[] =
 			NULL
 		},
 		&optimizer_analyze_root_partition,
-		false, NULL, NULL
+		true, NULL, NULL
```
Comment (bhuvnesh2703) [inline src/backend/utils/misc/guc_gp.c]: does it require any analyze test files to be updated?

File: src/backend/gpopt/config/CConfigParamMapping.cpp [hunk 2]
```diff
@@ -209,17 +209,17 @@ CConfigParamMapping::SConfigMappingElem CConfigParamMapping::m_elem[] =
...
 		},
 
 		{
-		EopttraceDonotCollectMissingStatsCols,
-		&optimizer_disable_missing_stats_collection,
+		EopttracePrintColsWithMissingStats,
+		&optimizer_print_missing_stats,
 		false, // m_fNegate
-		GPOS_WSZ_LIT("Disable collection of columns with missing statistics.")
+		GPOS_WSZ_LIT("Print columns with missing missing statistics.")
```
Comment (bhuvnesh2703) [inline src/backend/gpopt/config/CConfigParamMapping.cpp]: missing missing (2 times)

Comment (bhuvnesh2703) [inline src/backend/gpopt/config/CConfigParamMapping.cpp]: There are few more changes required in concourse pipeline and test cases
```
orinda ~/w/gpdb (streamline-orca-gucs ↩=)> ag optimizer_disable_missing_stats_collection
concourse/pipelines/pipeline.yml
1238:        BLDWRAP_POSTGRES_CONF_ADDONS: fsync=off optimizer_disable_missing_stats_collection=on
1246:        BLDWRAP_POSTGRES_CONF_ADDONS: fsync=off optimizer_disable_missing_stats_collection=on

src/test/regress/expected/bfv_dd.out
6:set optimizer_disable_missing_stats_collection=on;

src/test/regress/expected/bfv_dd_optimizer.out
6:set optimizer_disable_missing_stats_collection=on;
orinda ~/w/gpdb (streamline-orca-gucs ↩=)>
```

File: src/backend/utils/misc/guc_gp.c [hunk 3]
```diff
@@ -4457,7 +4455,7 @@ struct config_int ConfigureNamesInt_gp[] =
 	},
 
 	{
-		{"optimizer_cte_inlining_bound", PGC_USERSET, DEVELOPER_OPTIONS,
+		{"optimizer_cte_inlining_bound", PGC_USERSET, QUERY_TUNING_METHOD,
```
Comment (bhuvnesh2703) [inline src/backend/utils/misc/guc_gp.c]: There are other GUCs like optimizer_force_multistage_agg, optimizer_force_three_stage_scalar_dqa etc mentioned under `Optimizer Hints` section in guc.h which are set as DEVELOPER_OPTIONS, they will also fall in the QUERY_TUNING_METHOD category, isn't it?

File: src/test/regress/expected/gp_optimizer.out [hunk 4]
```diff
@@ -9073,6 +9073,7 @@ with x as (select * from foo_missing_stats) select count(*) from x x1, x x2 wher
 (1 row)
 
 set allow_system_table_mods="DML";
+set optimizer_print_missing_stats = on;
```
Comment (bhuvnesh2703) [inline src/test/regress/expected/gp_optimizer.out]: optimizer_print_missing_stats is true by default do we need it?

File: src/test/regress/sql/gp_optimizer.sql [hunk 5]
```diff
@@ -895,14 +895,15 @@ with x as (select * from foo_missing_stats) select count(*) from x x1, x x2 wher
 with x as (select * from foo_missing_stats) select count(*) from x x1, x x2 where x1.a = x2.b;
 
 set allow_system_table_mods="DML";
+set optimizer_print_missing_stats = on;
```
Comment (bhuvnesh2703) [inline src/test/regress/sql/gp_optimizer.sql]: trivial, optimizer_print_missing_stats is true by default do we need it?

File: src/test/regress/expected/bfv_aggregate.out [hunk 6]
```diff
@@ -177,14 +177,14 @@ insert into multi_stage_test select i, i%4 from generate_series(1,10) i;
 analyze multi_stage_test;
 -- TEST
 set optimizer_segments=2;
-set optimizer_prefer_multistage_agg = on;
+set optimizer_force_multistage_agg = on;
```
Comment (bhuvnesh2703) [inline src/test/regress/expected/bfv_aggregate.out]: trivial, this is on by default.

File: src/test/regress/expected/bfv_aggregate_optimizer.out [hunk 7]
```diff
@@ -176,14 +176,14 @@ insert into multi_stage_test select i, i%4 from generate_series(1,10) i;
 analyze multi_stage_test;
 -- TEST
 set optimizer_segments=2;
-set optimizer_prefer_multistage_agg = on;
+set optimizer_force_multistage_agg = on;
```
Comment (bhuvnesh2703) [inline src/test/regress/expected/bfv_aggregate_optimizer.out]: trivial, this is on by default.

File: src/test/regress/sql/bfv_dd.sql [hunk 8]
```diff
@@ -3,7 +3,6 @@
 --
 -- start_ignore
 set optimizer_log=on;
-set optimizer_disable_missing_stats_collection=on;
 -- end_ignore
```
Comment (bhuvnesh2703) [inline src/test/regress/sql/bfv_dd.sql]: should we have optimizer_print_missing_stats = off, here?

File: src/test/regress/sql/portals.sql [hunk 9]
```diff
@@ -1,7 +1,7 @@
+set optimizer_print_missing_stats = off;
 --
 -- Cursor regression tests
 --
-SET optimizer_disable_missing_stats_collection=true;
 BEGIN;
```
Comment (bhuvnesh2703) [inline src/test/regress/sql/portals.sql]: should we have optimizer_print_missing_stats = off, here?

File: src/test/regress/input/auth_constraint.source [hunk 10]
```diff
@@ -1,7 +1,6 @@
 --- By design, every boolean query (which doesn't error out) in this test must return true.
 --- For reference, 2011-08-30 was a Tuesday!
 ---
-set optimizer_disable_missing_stats_collection = on;
```
Comment (bhuvnesh2703) [inline src/test/regress/input/auth_constraint.source]: should we have optimizer_print_missing_stats = off, here?

File: src/test/regress/input/external_table.source [hunk 11]
```diff
@@ -25,7 +25,6 @@
 -- s/cat\: cannot open (.*)$/cat\: $1\: NO SUCH FILE/
 --
 -- end_matchsubs
-set optimizer_disable_missing_stats_collection = on;
```
Comment (bhuvnesh2703) [inline src/test/regress/input/external_table.source]: should we have optimizer_print_missing_stats = off, here?

File: src/test/regress/output/auth_constraint.source [hunk 12]
```diff
@@ -1,7 +1,6 @@
 --- By design, every boolean query (which doesn't error out) in this test must return true.
 --- For reference, 2011-08-30 was a Tuesday!
 ---
-set optimizer_disable_missing_stats_collection = on;
```
Comment (bhuvnesh2703) [inline src/test/regress/output/auth_constraint.source]: should we have optimizer_print_missing_stats = off, here?

File: src/test/regress/expected/portals.out [hunk 13]
```diff
@@ -1,7 +1,7 @@
+set optimizer_print_missing_stats = off;
 --
 -- Cursor regression tests
 --
-SET optimizer_disable_missing_stats_collection=true;
```
Comment (bhuvnesh2703) [inline src/test/regress/expected/portals.out]: should we have optimizer_print_missing_stats = off, here?

File: src/backend/utils/misc/guc_gp.c [hunk 14]
```diff
@@ -2518,12 +2518,12 @@ struct config_bool ConfigureNamesBool_gp[] =
 	},
 
 	{
-		{"optimizer_disable_missing_stats_collection", PGC_USERSET, LOGGING_WHAT,
-			gettext_noop("Disable collecting of columns with missing statistics."),
+		{"optimizer_print_missing_stats", PGC_USERSET, LOGGING_WHAT,
+			gettext_noop("Print columns with missing missing statistics."),
```
Comment (danielgustafsson) [inline src/backend/utils/misc/guc_gp.c]: Same double "missing" as @bhuvnesh2703 commented on above.

File: src/include/utils/guc.h [hunk 15]
```diff
@@ -403,14 +406,13 @@ extern bool	optimizer_print_expression_properties;
...
+
+/* array of xforms disable flags */
+extern bool optimizer_xforms[OPTIMIZER_XFORMS_COUNT];
 extern char *optimizer_search_strategy_path;
-extern bool optimizer_extract_dxl_stats;
-extern bool optimizer_extract_dxl_stats_all_nodes;
-extern bool optimizer_disable_missing_stats_collection;
-extern bool optimizer_dpe_stats;
+
+/* GUCs to tell Optimizer to enable a physical operator*/
```
Comment (danielgustafsson) [inline src/include/utils/guc.h]: Space between "r" and end of comment marker

File: src/backend/utils/misc/guc_gp.c (hunk 1 above)
Comment (vraghavan78) [inline src/backend/utils/misc/guc_gp.c]: no. I think the analyze test does not make any assumptions on the default GUC value

File: src/test/regress/sql/portals.sql (hunk 9 above)
Comment (vraghavan78) [inline src/test/regress/sql/portals.sql]: passing with out this guc. I do not want to litter these gucs in tests when they are no needed.

File: src/test/regress/sql/bfv_dd.sql (hunk 8 above)
Comment (vraghavan78) [inline src/test/regress/sql/bfv_dd.sql]: passing with out this guc. I do not want to litter these gucs in tests when they are no needed.

File: src/test/regress/input/external_table.source (hunk 11 above)
Comment (vraghavan78) [inline src/test/regress/input/external_table.source]: same as above

File: src/test/regress/output/auth_constraint.source (hunk 12 above)
Comment (vraghavan78) [inline src/test/regress/output/auth_constraint.source]: same here

File: src/test/regress/expected/portals.out (hunk 13 above)
Comment (vraghavan78) [inline src/test/regress/expected/portals.out]: same here

File: src/backend/utils/misc/guc_gp.c [hunk 16]
```diff
@@ -4884,7 +4882,7 @@ struct config_string ConfigureNamesString_gp[] =
 	},
 
 	{
-		{"optimizer_cost_model", PGC_USERSET, LOGGING_WHEN,
+		{"optimizer_cost_model", PGC_USERSET, DEVELOPER_OPTIONS,
```
Comment (vraghavan78) [inline src/backend/utils/misc/guc_gp.c]: no, we are no longer supporting legacy cost model. We may even get rid of this guc altogether

File: src/test/regress/input/auth_constraint.source (hunk 10 above)
Comment (vraghavan78) [inline src/test/regress/input/auth_constraint.source]: same as above

File: src/backend/utils/misc/guc_gp.c (hunk 16 above)
Comment (danielgustafsson) [inline src/backend/utils/misc/guc_gp.c]: :+1: on removing the guc and leftover code completely while we're at it.

File: src/test/regress/sql/portals.sql (hunk 9 above)
Comment (danielgustafsson) [inline src/test/regress/sql/portals.sql]: +1

File: src/backend/gpopt/config/CConfigParamMapping.cpp (hunk 2 above)
Comment (bhuvnesh2703) [inline src/backend/gpopt/config/CConfigParamMapping.cpp]: fixed the above and pushed

File: src/include/utils/guc.h [hunk 17]
```diff
@@ -422,55 +424,67 @@ extern bool optimizer_enable_materialize;
...
 extern bool optimizer_enable_broadcast_nestloop_outer_child;
-extern bool optimizer_enforce_subplans;
 extern bool optimizer_enable_assert_maxonerow;
+extern bool optimizer_enable_constant_expression_evaluation;
+extern bool optimizer_enable_bitmapscan;
+extern bool optimizer_enable_outerjoin_to_unionall_rewrite;
+extern bool optimizer_enable_ctas;
+extern bool optimizer_enable_partial_index;
+extern bool optimizer_enable_dml_triggers;
+extern bool	optimizer_enable_dml_constraints;
```
Comment (hsyuan) [inline src/include/utils/guc.h]: indent

File: src/backend/gpopt/config/CConfigParamMapping.cpp [hunk 18]
```diff
@@ -48,10 +48,10 @@ CConfigParamMapping::SConfigMappingElem CConfigParamMapping::m_elem[] =
 		},
 
 		{
-		EopttraceDisablePrintXformRes,
-		&optimizer_disable_xform_result_printing,
+		EopttracePrintXformRes,
```
Comment (oarap) [inline src/backend/gpopt/config/CConfigParamMapping.cpp]: Can we have one to one mapping from guc name to the enum with no abbreviations such as `results` to `Res`.

`optimizer_aaa_bbb_ccc` to `EopttraceAaaBbbCcc`

File: src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp [hunk 19]
```diff
@@ -4401,7 +4401,7 @@ CTranslatorDXLToPlStmt::PlDirectDispatchSegIds
 	CDXLDirectDispatchInfo *pdxlddinfo
 	)
 {
-	if (!optimizer_direct_dispatch || NULL == pdxlddinfo)
+	if (!optimizer_enable_direct_dispatch || NULL == pdxlddinfo)
```
Comment (oarap) [inline src/backend/gpopt/translate/CTranslatorDXLToPlStmt.cpp]: So we git rid of `disable` but like `enable` and add even more `enable`. I think we should remove `enable`s as well

File: src/backend/utils/misc/guc_gp.c [hunk 20]
```diff
@@ -454,20 +454,22 @@ bool		gp_log_dynamic_partition_pruning = false;
 bool		gp_cte_sharing = false;
 bool		gp_enable_relsize_collection = false;
 
-/* ORCA related gucs */
-bool		optimizer_control = true;
+/* Optimizer related gucs */
 bool		optimizer;
 bool		optimizer_log;
+int			optimizer_log_failure;
+bool		optimizer_control = true;
```
Comment (oarap) [inline src/backend/utils/misc/guc_gp.c]: this guc is useless

File: src/backend/utils/misc/guc_gp.c [hunk 21]
```diff
@@ -454,20 +454,22 @@ bool		gp_log_dynamic_partition_pruning = false;
...
 bool		optimizer_trace_fallback;
 bool		optimizer_partition_selection_log;
 bool		optimizer_minidump;
 int			optimizer_cost_model;
```
Comment (oarap) [inline src/backend/utils/misc/guc_gp.c]: indentation

File: src/backend/utils/misc/guc_gp.c [hunk 22]
```diff
@@ -454,20 +454,22 @@ bool		gp_log_dynamic_partition_pruning = false;
...
+bool		optimizer_metadata_caching;
+int			optimizer_mdcache_size;
```
Comment (oarap) [inline src/backend/utils/misc/guc_gp.c]: indentation

**Comment by vraghavan78** [comment]
@oarap I have update the traceflag names to match the guc name. Please take another look. I do not want to remove enable, since it follows planner's enable style for nodes. I removed it for tuning and internal Orca knobs.

**Comment by vraghavan78** [comment]
Merged Thanks!

**Pull request closed without merging**